### PR TITLE
(docs) - Add page for more detailed auth docs

### PR DESCRIPTION
**Pull request by kadikraman**
🔐

### CODE REVIEW

**Review by JoviDeCroock — APPROVED** [review]
This is so detailled, awesome! 💯

File: docs/advanced/authentication.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+const didAuthError = ({ error }) => {
+  return error.graphQLErrors.some(
+    e => error.response.status === 401,
+  );
+},
+```
+
+Then `didAuthError` returns `true`, it will trigger the exchange to trigger the logic for asking for reauthentication via `getAuth`.
+
+### Configuring `getAuth` (after `didAuthError` returns `true`)
```
Comment (kitten) [inline docs/advanced/authentication.md:219]: If "after `didAuthError` returns `true` is important information and there's no corresponding section that for instance says `false` instead, let's move this into the text below as a full explanation?

> `getAuth` will be triggered to refresh the authentication state if `didAuthError` returned `true` previously, due to at least one GraphQL request erroring out with an authentication error, as we defined previously.

File: docs/advanced/authentication.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    localStorage.setItem('refreshToken', result.data.refreshLogin.refreshToken);
+
+    return {
+      token: result.data.refreshLogin.token,
+      refreshToken: result.data.refreshLogin.refreshToken,
+    };
+  }
+
+  localStorage.clear();
+  logout();
```
Comment (kitten) [inline docs/advanced/authentication.md:270]:
```suggestion
  // This is where auth has gone wrong and we need to clean up and redirect to a login page
  localStorage.clear();
  logout();
```

Just adding a comment here since this isn't reusable for the user's app

File: docs/advanced/authentication.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+  logout();
+
+  return null;
+}
+```
+
+Here, `logout()` should clear your persisted storage of any tokens and set the `isLoggedIn` property as described above to `false`.
+
+If you do have a token refresh endpoint, and a refresh token, you can attempt to get a new token for the user:
```
Comment (kitten) [inline docs/advanced/authentication.md:242]:
```suggestion
Here, `logout()` is a placeholder that is called when we got an error, so that we can redirect to a login page again and clear our tokens from local storage or otherwise.

If we had a way to refresh our token using a refresh token, we can attempt to get a new token for the user first:
```

Just rewriting to keep the first-person plural consistent

File: docs/advanced/authentication.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+  }
+
+  localStorage.clear();
+  logout();
+
+  return null;
+}
+```
+
+Here we use the special mutate function provided by the auth exchange to do the token refresh. If your auth is not handled via graphQL, you can
```
Comment (kitten) [inline docs/advanced/authentication.md:276]:
```suggestion
Here we use the special mutate function provided by the auth exchange to do the token refresh. If your auth is not handled via GraphQL but a REST endpoint, you can
```

File: docs/advanced/authentication.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+use `fetch` in this function instead of a mutation. All other requests will be paused while `getAuth` returns, so you won't get multiple auth failures
```
Comment (kitten) [inline docs/advanced/authentication.md:277]:
```suggestion
use `fetch` in this function instead of a mutation. All other requests will be paused while `getAuth` returns, so we never have to handle multiple auth failures
```

File: docs/advanced/authentication.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+at the same time.
+
+### Configuring `willAuthError`
+
+`willAuthError` is an optional parameter and is run _before_ a network request is made. You can use it to trigger the logic in
+`getAuth` without the need to get a graphql error first. For example, you could use this indicate that the auth will fail because the JWT is invalid:
```
Comment (kitten) [inline docs/advanced/authentication.md:283]:
```suggestion
`willAuthError` is an optional parameter and is run _before_ a network request is made. We can use it to trigger the logic in
`getAuth` without the need to send a request and get a GraphQL Error back. For example, we can use this to predict that the authentication will fail because our JWT is invalid already:
```

File: docs/advanced/authentication.md [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+const willAuthError = ({ authState }) => {
+  if (!authState || /* JWT is expired */) return true;
+  return false;
+}
+```
+
+## Handling Logout with the Error Exchange
+
+It is also possible to handle your logout logic in the error exchange instead of in the auth exchange. To do this, you'll need to add the
+`errorExchange` to the exchanges array, _before_ the `authExchange`. The order is very important here:
```
Comment (kitten) [inline docs/advanced/authentication.md:295]:
```suggestion
We can also handle authentication errors in an `errorExchange` instead of the `authExchange`. To do this, we'll need to add the
`errorExchange` to the exchanges array, _before_ the `authExchange`. The order is very important here:
```

File: docs/advanced/authentication.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+import { createClient, dedupExchange, cacheExchange, fetchExchange, errorExchange } from 'urql';
+import { authExchange } from '@urql/exchange-auth';
+
+const client = createClient({
+  url: '/graphql',
+  exchanges: [
+    dedupExchange,
+    cacheExchange,
+    errorExchange({
+      onError: ({ error }) => {
```
Comment (kitten) [inline docs/advanced/authentication.md:307]:
```suggestion
      onError: (error) => {
```

File: docs/advanced/authentication.md [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    }),
+    fetchExchange,
+  ],
+});
+```
+
+The `errorExchange` will only receive an auth error when the auth exchange as tried and failed to handle it. This means we have
+either failed to refresh the token, or there is no token refresh functionality. If you get an auth error in the `errorExchange` (as defined in
+the `didAuthError` configuration section above), then you can be confident that it is an auth error, the `authExchange` isn't able to recover
+from it, and the user should be logged out.
```
Comment (kitten) [inline docs/advanced/authentication.md:328]:
```suggestion
The `errorExchange` will only receive an auth error when the auth exchange has already tried and failed to handle it. This means we have
either failed to refresh the token, or there is no token refresh functionality. If we receive an auth error in the `errorExchange` (as defined in
the `didAuthError` configuration section above), then we can be confident that it is an auth error that the `authExchange` isn't able to recover
from, and the user should be logged out.
```

File: docs/advanced/authentication.md (hunk 1 above)
Comment (kadikraman) [reply docs/advanced/authentication.md:219]: The two sections are "Configuring `getAuth` (fetch from storage)" and "Configuring `getAuth` (after `didAuthError` returns `true`)".

I could rephrase it to "after an auth error has happened"?

**Review by kitten — APPROVED** [review]
